akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -635,9 +611,18 @@ private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Mo
       s"The top level module cannot be materialized because it has unconnected ports: ${(topLevel.inPorts ++ topLevel.outPorts).mkString(", ")}")
     try materializeModule(topLevel, topLevel.attributes)
     catch {
-      case NonFatal(e) ⇒
-        panic(e)
-        throw e
+      case NonFatal(cause) ⇒
+        // Cancels all intermediate Publishers and fails all intermediate Subscribers.
+        // (This is an attempt to clean up after an exception during materialization)
+        val panicError = new MaterializationPanic(cause)
+        val errorPublisher = new ErrorPublisher(cause, "")
```

Comment:
what is not intuitive? I cannot really say anything to these comments, because I don't see what you have problems with. 

I try to explain what happens here though. If there is an exception during materialization (note, these are not "in-stream" exceptions, but exceptions when wiring up the graph), then we need to do our best to clean up. Unfortunately at this point we have a partially materialized graph, with potentially dangling Publishers and Subscribers, some of which might be 3rd party library entities actually.

This means that we just take all the Publishers/Subscriber that we know of, and try to shut them down, even if this means a Spec violation (there is basically nothing else we can do here, this is a best effort approach).

As for the stacks, these implement the scoped materialization, that is necessary to avoid copying graphs when you import them. These implement something similar to variable scopes in a language like Scala.